torch/_inductor/codegen/cpp.py

```diff
@@ -3714,26 +3702,54 @@ def is_lowp_fp_scheduler(self, scheduler_node: SchedulerNode):
 
     def legalize_lowp_fp_dtype_loopbody(self, loop_body: LoopBody):
         def add_to_dtype(sub_graph: torch.fx.Graph):
-            def is_lowp_fp_load(node: torch.fx.Node):
-                if node.target not in ["load"]:
-                    return False
-                assert len(node.args) == 3
-                load_dtype = V.graph.get_dtype(node.args[1])  # type: ignore[arg-type]
-                return load_dtype in DTYPE_LOWP_FP
-
-            def is_lowp_fp_store(node: torch.fx.Node):
-                if node.target != "store":
+            def get_input_dtype(node: torch.fx.Node):
+                """Get input dtype for nodes that may consumes lowp fp dt"""
+                if node.target == "store":
+                    return V.graph.get_dtype(node.args[1])
+                elif node.target == "to_dtype_bitcast":
+                    return node.args[-1]
+                elif node.target == "to_dtype":
+                    return node.kwargs.get("src_dtype", None)
+                else:
+                    return None
+
+            def get_output_dtype(node: torch.fx.Node):
+                """Get output dtype for nodes that may produce lowp fp dt"""
+                if node.target == "load":
+                    assert len(node.args) == 3
+                    return V.graph.get_dtype(node.args[1])
+                elif node.target == "constant":
+                    return node.args[-1]
+                elif node.target == "to_dtype":
+                    return node.args[-1]
+                elif node.target == "to_dtype_bitcast":
+                    return node.args[2]
+                else:
+                    return None
+
+            def is_lowp_fp_source(node: torch.fx.Node, dt: torch.dtype):
+                """If the node produce the lowp fp dt as output without casting to high precision"""
+                return get_output_dtype(node) == dt
+
+            def is_lowp_fp_sink(node: torch.fx.Node, dt: torch.dtype):
+                """If the node can accepts the lowp fp dt as input without casting to high precision"""
```

Comment:
"""Check if the given node accept input with expected low precision floating point data type."""

Reply:
Updated in a63deaf